### PR TITLE
[Refine] Normalize device handling in state dicts and more

### PR DESCRIPTION
**Pull request by 0xWJ**

### CODE REVIEW

File: nnscaler/parallel.py [hunk 1]
```diff
@@ -1790,11 +1790,8 @@ def merge_state_dicts(
     Please Note:
 
     We don't garantee the devices of tensors are the same in the merged state dict.
-    You can assume the device of the tensors in the merged state dict can be one of the following:
-
-    1. the current device when running this function
-    2. the current cuda device when running this function
-    3. the device of the tensor in the original state dict
+    You can assume the device of the tensors in the merged state dict
+    can be 'cpu' or the device of the tensor in the original state dict.
```
Comment (zyeric) [inline nnscaler/parallel.py:1794]: you mean after this PR, it is possible that some tensors are still on GPU?

Comment (0xWJ) [reply nnscaler/parallel.py:1794]: In current implementation, we just take the tensor from input state dicts for non-parallel modules. 

For parallel modules, the current implementation will create new tensors from cpu, and copy/merge the tensors from input state dicts to it. (this may be optimized later as we can avoid copying for replicated tensors.)

So in summary, the devices of the tensors in output state dicts can be either 'cpu' or the device in original state dict.

File: nnscaler/utils.py [hunk 2]
```diff
@@ -884,20 +891,29 @@ def broadcast_mixed_data(
     torch.distributed.broadcast_object_list(sent, src=src_rank, group=group)
     skeleton, meta_tensors = sent[0]
```
Comment (lynex) [inline nnscaler/utils.py:892]: more clear if move under the below if statement

File: nnscaler/utils.py [hunk 3]
```diff
@@ -884,20 +891,29 @@ def broadcast_mixed_data(
...
-        tensors = [torch.empty_like(mt, device='cuda') for mt in meta_tensors]
-    else:
-        # make sure tensors are in cuda
-        tensors = [t.cuda() for t in tensors]
+        tensors = [None] * len(meta_tensors)
 
     # broadcast tensor data
     for i in range(len(tensors)):
-        torch.distributed.broadcast(tensors[i], src=src_rank, group=group)
+        if rank != src_rank:
```
Comment (lynex) [inline nnscaler/utils.py:898]: personally prefer aligning if-statement, e.g., Ln 879 checking '== src_rank...else' but here '!= src_rank...else'

File: nnscaler/parallel.py [hunk 4]
```diff
@@ -2314,13 +2311,11 @@ def _extract_new_state(
...
                 if sliced_new_val[key].numel() < zero3_info.chunk_size:
                     # padding if needed
-                    sliced_new_val[key] = torch.cat(
-                        [sliced_new_val[key],
-                         torch.zeros(
-                             zero3_info.chunk_size - sliced_new_val[key].numel(),
-                             dtype=sliced_new_val[key].dtype,
-                             device=sliced_new_val[key].device
-                         )], dim=0
+                    sliced_new_val[key] = torch.nn.functional.pad(
```
Comment (lynex) [inline nnscaler/parallel.py:2321]: 1. why changing from cat to pad?
2. datatypes will follow sliced_new_val[key] automatically, right?

**Review by lynex — COMMENTED** [review]
commented

File: nnscaler/utils.py (hunk 3 above)
Comment (0xWJ) [reply nnscaler/utils.py:898]: Agree, but not a big deal.

File: nnscaler/utils.py (hunk 2 above)
Comment (0xWJ) [reply nnscaler/utils.py:892]: I like the current flow. I am trying to put logic out of if-branches.

File: nnscaler/parallel.py (hunk 4 above)
Comment (0xWJ) [reply nnscaler/parallel.py:2321]: 1. reduce memory allocation. (cat will have 1 more allocation)
2. yes

File: nnscaler/parallel.py (hunk 1 above)
Comment (zyeric) [reply nnscaler/parallel.py:1794]: I see, can you put your explanation in the comments, thanks!

**Review by zyeric — APPROVED** [review]

**Review by lynex — APPROVED** [review]

File: nnscaler/parallel.py (hunk 1 above)
Comment (0xWJ) [reply nnscaler/parallel.py:1794]: added
